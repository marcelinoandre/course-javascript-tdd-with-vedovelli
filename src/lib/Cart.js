import find from 'lodash/find';
import remove from 'lodash/remove';
import Dinero from 'dinero.js';

const Money = Dinero;

Money.defaultCurrency = 'BRL';
Money.defaultPrecision = 2;

const calculatePercentageDiscount = (amount, item) => {
  if (item.condition?.percentage && item.quantity > item.condition.minimum) {
    return amount.percentage(item.condition.percentage);
  }
  return Money({ amount: 0 });
};

export class Cart {
  items = [];

  add(item) {
    const itemTofind = { product: item.product };
    //TODO: remover lodash e fazer com javascript puro
    if (find(this.items, itemTofind)) remove(this.items, itemTofind);
    this.items.push(item);
  }

  getTotal() {
    return this.items.reduce((accumulator, item) => {
      const amount = Money({ amount: item.quantity * item.product.price });
      const  discount = calculatePercentageDiscount(amount, item)

      return accumulator.add(amount).subtract(discount);
    }, Money({ amount: 0 }));
  }

  remove(product) {
    remove(this.items, { product });

    return this.getTotal();
  }

  summary() {
    const total = this.getTotal().getAmount();
    const items = this.items;

    return { total, items };
  }

  checkout() {
    const total = this.getTotal();
    const items = this.items;

    this.items = [];

    return { total, items };
  }
}
