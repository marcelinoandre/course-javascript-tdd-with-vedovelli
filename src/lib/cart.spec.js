import { Cart } from './Cart';

describe('Cart', () => {
  let cart;
  let product = {
    title: 'Adidas running shoes - men 41',
    price: 35388 //353.88 | R$ 383,88
  };
  let product2 = {
    title: 'Adidas running shoes - women 39',
    price: 41872 //418.72 | R$ 418,72
  };

  beforeEach(() => {
    cart = new Cart();
  });
  describe('getTotal()', () => {
    it('should return 0 when getTotal() is executed in a newly created instance', () => {
      expect(cart.getTotal().getAmount()).toEqual(0);
    });

    it('should multiply quantity and price and receive the total amount', () => {
      const item = {
        product,
        quantity: 2 // 70776
      };

      cart.add(item);
      expect(cart.getTotal().getAmount()).toEqual(70776);
    });

    it('should ensure no more than on product exists at a time', () => {
      cart.add({
        product,
        quantity: 2
      });
      cart.add({
        product,
        quantity: 1
      });

      expect(cart.getTotal().getAmount()).toEqual(35388);
    });

    it('should update total when a product gets included and then removed', () => {
      cart.add({
        product,
        quantity: 2
      });
      cart.add({
        product: product2,
        quantity: 1
      });

      cart.remove(product);

      expect(cart.getTotal().getAmount()).toEqual(41872);
    });
  });

  describe('checkout()', () => {
    it('should return an object with the total and the list of items', () => {
      cart.add({
        product,
        quantity: 5
      });

      cart.add({
        product: product2,
        quantity: 3
      });
      expect(cart.summary()).toMatchSnapshot();
      expect(cart.getTotal().getAmount()).toBeGreaterThan(0);
    });

    it('should return an object with the total and the list of items', () => {
      cart.add({
        product,
        quantity: 5
      });

      cart.add({
        product: product2,
        quantity: 3
      });
      expect(cart.checkout()).toMatchSnapshot();
    });

    it('should reset the cart when checkout() is called', () => {
      cart.add({
        product: product2,
        quantity: 3
      });
      expect(cart.getTotal().getAmount()).toEqual(125616);
    });
  });

  describe('special conditions of discounts', () => {
    it('should apply percentage discount quantity above minimum is passed', () => {
      const condition = {
        percentage: 30,
        minimum: 2
      };
      cart.add({
        product,
        condition,
        quantity: 3
      });
      expect(cart.getTotal().getAmount()).toEqual(74315);
    });

    it('should apply quantity discount for even quantities', () => {});
  });
});
