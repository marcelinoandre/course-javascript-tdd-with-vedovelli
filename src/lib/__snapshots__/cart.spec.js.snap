// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shoes - men 41",
      },
      "quantity": 5,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas running shoes - women 39",
      },
      "quantity": 3,
    },
  ],
  "total": 302556,
}
`;

exports[`Cart checkout() should return an object with the total and the list of items 2`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shoes - men 41",
      },
      "quantity": 5,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas running shoes - women 39",
      },
      "quantity": 3,
    },
  ],
  "total": Object {
    "amount": 302556,
    "currency": "BRL",
    "precision": 2,
  },
}
`;
